vault_renewal.go

```diff
@@ -2,129 +2,141 @@ package main
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"log"
 
 	vault "github.com/hashicorp/vault/api"
 )
 
 // Once you've set the token for your Vault client, you will need to
-// periodically renew its lease.
+// periodically renew it. Likewise, the database credentials lease will expire
+// at some point and also needs to be renewed periodically.
 //
-// A function like this should be run as a goroutine to avoid blocking.
-//
-// Production applications may also wish to be more tolerant of failures and
-// retry rather than exiting.
+// A function like this one should be run as a goroutine to avoid blocking.
+// Production applications may also need to be more tolerant of failures and
+// retry on errors rather than exiting.
 //
 // Additionally, enterprise Vault users should be aware that due to eventual
 // consistency, the API may return unexpected errors when running Vault with
 // performance standbys or performance replication, despite the client having
-// a freshly renewed token. See https://www.vaultproject.io/docs/enterprise/consistency#vault-1-7-mitigations
-// for several ways to mitigate this which are outside the scope of this code sample.
-func (v *Vault) RenewLoginPeriodically(ctx context.Context, authToken *vault.Secret) {
-	/* */ log.Println("auth token renew / login loop: begin")
-	defer log.Println("auth token renew / login loop: end")
+// a freshly renewed token. See the link below for several ways to mitigate
+// this which are outside the scope of this code sample.
+//
+// ref: https://www.vaultproject.io/docs/enterprise/consistency#vault-1-7-mitigations
+func (v *Vault) PeriodicallyRenewSecrets(
+	ctx context.Context,
+	authToken *vault.Secret,
+	databaseCredentialsLease *vault.Secret,
+	databaseReconnectFunc func(ctx context.Context, credentials DatabaseCredentials) error,
+) {
+	/* */ log.Println("renew / recreate secrets loop: begin")
+	defer log.Println("renew / recreate secrets loop: end")
 
 	currentAuthToken := authToken
+	currentDatabaseCredentialsLease := databaseCredentialsLease
 
 	for {
-		if err := v.renewUntilMaxTTL(ctx, currentAuthToken, "auth token"); err != nil {
-			// break out when shutdown is requested
-			if errors.Is(err, context.Canceled) {
-				return
-			}
-
-			log.Fatalf("auth token renew error: %v", err) // simplified error handling
-		}
-
-		// the auth token's lease has expired and needs to be renewed
-		t, err := v.login(ctx)
+		r, err := v.renewLeases(ctx, currentAuthToken, currentDatabaseCredentialsLease)
```

Comment:
nit: I'm really used to seeing r as the variable that represents a *http.Request (like in handler code), so I try to avoid using it in other contexts to reduce confusion. 

Would it make more sense for renewLeases to return a (bool, error) and name the var `renewed`? Or if it has to be the uint8 for some reason, maybe just `ren` or `res` or something....

Reply:
I'll go with `result` to be clear

Reply:
actually, I like your `renewed` suggestion more on a second thought 😄